code/modules/mob/living/basic/minebots/minebot_ai.dm

```diff
@@ -0,0 +1,306 @@
+	. = ..()
+	minimum_distance = controller.blackboard[BB_MINIMUM_SHOOTING_DISTANCE] ?  controller.blackboard[BB_MINIMUM_SHOOTING_DISTANCE] : initial(minimum_distance)
+	var/atom/target = controller.blackboard[target_key]
+	if(QDELETED(target))
+		return AI_BEHAVIOR_DELAY | AI_BEHAVIOR_FAILED
+	var/mob/living/living_pawn = controller.pawn
+	if(get_dist(living_pawn, target) <= minimum_distance)
+		return AI_BEHAVIOR_DELAY | AI_BEHAVIOR_SUCCEEDED
+
+///mine walls if we are on automated mining mode
+/datum/ai_planning_subtree/minebot_mining/select_behaviors(datum/ai_controller/controller, seconds_per_tick)
+	if(!controller.blackboard[BB_AUTOMATED_MINING])
+		return
+	if(controller.blackboard_key_exists(BB_MINEBOT_GIBTONITE_RUN))
+		return
+	if(controller.blackboard_key_exists(BB_TARGET_MINERAL_TURF))
+		controller.queue_behavior(/datum/ai_behavior/minebot_mine_turf, BB_TARGET_MINERAL_TURF)
+		return SUBTREE_RETURN_FINISH_PLANNING
+	controller.queue_behavior(/datum/ai_behavior/find_mineral_wall/minebot, BB_TARGET_MINERAL_TURF)
+
+/datum/ai_behavior/find_mineral_wall/minebot
+
+/datum/ai_behavior/find_mineral_wall/minebot/check_if_mineable(datum/ai_controller/controller, turf/target_wall)
+	var/list/forbidden_turfs = controller.blackboard[BB_BLACKLIST_MINERAL_TURFS]
+	var/turf/previous_unreachable_wall = controller.blackboard[BB_PREVIOUS_UNREACHABLE_WALL]
+	if(is_type_in_list(target_wall, forbidden_turfs) || target_wall == previous_unreachable_wall)
+		return FALSE
+	controller.clear_blackboard_key(BB_PREVIOUS_UNREACHABLE_WALL)
+	return ..()
+
+/datum/ai_behavior/minebot_mine_turf
+	behavior_flags = AI_BEHAVIOR_REQUIRE_MOVEMENT | AI_BEHAVIOR_CAN_PLAN_DURING_EXECUTION
+	required_distance = 2
+	action_cooldown = 3 SECONDS
+
+/datum/ai_behavior/minebot_mine_turf/setup(datum/ai_controller/controller, target_key)
+	. = ..()
+	var/turf/target = controller.blackboard[target_key]
+	if(isnull(target))
+		return FALSE
+	set_movement_target(controller, target)
+
+/datum/ai_behavior/minebot_mine_turf/perform(seconds_per_tick, datum/ai_controller/controller, target_key)
+	var/mob/living/basic/living_pawn = controller.pawn
+	var/turf/target = controller.blackboard[target_key]
+
+	if(QDELETED(target))
+		return AI_BEHAVIOR_DELAY | AI_BEHAVIOR_FAILED
+
+	if(check_obstacles_in_path(controller, target))
+		return AI_BEHAVIOR_DELAY | AI_BEHAVIOR_FAILED
+
+	var/old_intent = living_pawn.a_intent
+	if(living_pawn.a_intent != INTENT_HARM)
+		living_pawn.a_intent = INTENT_HARM
+
+	living_pawn.face_atom(target)
+
+	RegisterSignal(target, COMSIG_MINE_EXPOSE_GIBTONITE, PROC_REF(on_mine_expose_gibtonite))
+
+	living_pawn.RangedAttack(target)
+	living_pawn.a_intent = old_intent
+
+	return AI_BEHAVIOR_DELAY | AI_BEHAVIOR_SUCCEEDED
+
+/datum/ai_behavior/minebot_mine_turf/proc/on_mine_expose_gibtonite(datum/source, mob/living/instigator)
+	SIGNAL_HANDLER // COMSIG_MINE_EXPOSE_GIBTONITE
+
+	instigator.emote("me", EMOTE_VISIBLE, "yelps!")
```

Comment:
same here with the custom emote bit

Reply:
as above, custom_emote can potentially sleep so it's not suitable for signal handlers